README.md
# EventSource client for EventMachine

See the specification: https://html.spec.whatwg.org/multipage/server-sent-events.html

## Install

Install with Rubygems:

    gem install em-eventsource

If you use bundler, add it to your Gemfile:

    gem "em-eventsource", "~> 0.3.0"

## Usage

Basic usage:

```ruby
require "em-eventsource"
EM.run do
  source = EventMachine::EventSource.new("http://example.com/streaming")
  source.message do |message|
    puts "new message #{message}"
  end
  source.start # Start listening
end
```

Listening specific event name:

```ruby
source.on "eventname" do |message|
  puts "eventname #{message}"
end
```

Handle error:

```ruby
source.error do |error|
  puts "error #{error}"
end
```

Handle open stream:

```ruby
source.open do
  puts "opened"
end
```

Close the stream:

```ruby
source.close
```

Current status of the connection:

```ruby
# Can be:
# - EM::EventSource::CLOSED
# - EM::EventSource::CONNECTING
# - EM::EventSource::OPEN
source.ready_state
```

Override the default retry value (if the connection is lost):

```ruby
source.retry = 5 # in seconds (default 3)
```

Get Last-Event-Id value:

```ruby
source.last_event_id
```

Attach middleware:

```ruby
source.use EM::Middleware::JSONResponse
```

Set the inactivity timeout. Set to 0 to disable the timeout.

```ruby
source.inactivity_timeout = 120 # in seconds (default: 60).
```

## Licence

MIT License

Copyright (C) 2020 François de Metz

Copyright (C) 2011 af83


CHANGELOG.md
# em-eventsource changelog

## 0.3.2

- Add TLS hostname verification by default [[#19](https://github.com/francois2metz/em-eventsource/pull/19)]

## 0.3.1

- Force UTF-8 encoding on incoming events [[#16](https://github.com/francois2metz/em-eventsource/pull/16)]

## 0.3.0

- Handle retry values sent by the server as milliseconds [[#14](https://github.com/francois2metz/em-eventsource/issues/14)]


spec/spec_helper.rb
require "em-eventsource"

module EventMachine
  class MockHttpRequest
    attr_reader :url, :get_args, :middlewares, :opts, :response_header, :closed

    def initialize(url, opts={})
      @url = url
      @streams = []
      @errors = []
      @callbacks = []
      @headers = []
      @middlewares = []
      @opts = opts
      @response_header = HttpResponseHeader.new({})
      @closed = false
    end

    def get(*args)
      @get_args = args
      self
    end

    def stream(&block)
      @streams << block
    end

    def errback(&block)
      @errors << block
    end

    def callback(&block)
      @callbacks << block
    end

    def use(*args, &block)
      @middlewares << [*args, block]
    end

    def headers(&block)
      @headers << block
    end

    def stream_data(data)
      @streams.each { |stream| stream.call(data) }
    end

    def call_errback
      @errors.each { |error| error.call() }
    end

    def call_callback
      @callbacks.each { |callback| callback.call("chunk of data") }
    end

    def call_headers(headers)
      @response_header = headers
      @headers.each { |header| header.call(headers) }
    end

    def close(reason="")
      @closed = true
    end
  end

  class HttpRequest
    def self.new(url, opts={})
      EM::MockHttpRequest.new(url, opts)
    end
  end
end


spec/em-eventsource_spec.rb
#!/usr/bin/env ruby
# encoding: utf-8

require File.expand_path(File.dirname(__FILE__) + '/spec_helper')
require "minitest/autorun"
require "minitest-spec-context"

describe EventMachine::EventSource do

  def start_source(url="http://example.com/streaming", query={}, headers={})
    EM.run do
      source = EventMachine::EventSource.new(url, query, headers)
      source.start
      req = source.instance_variable_get "@req"
      yield source, req if block_given?
    end
  end

  def create_response_headers(status, content_type="", other={})
    headers = EM::HttpResponseHeader.new
    headers.http_status = status
    headers['CONTENT_TYPE'] = content_type
    headers.merge!(other)
  end

  it "connect to the good server" do
    start_source do |source, req|
      _(source.ready_state).must_equal EM::EventSource::CONNECTING
      _(source.url).must_equal "http://example.com/streaming"
      _(req.url).must_equal "http://example.com/streaming"
      _(req.opts[:inactivity_timeout]).must_equal 60
      _(req.get_args[0]).must_equal({ query: {},
                                   head: {"Cache-Control" => "no-cache",
                                          "Accept" => "text/event-stream"},
                                   keepalive: true })
      EM.stop
    end
  end

  it "connect to the good server with query and headers" do
    start_source "http://example.net/streaming", {:chuck => "norris"}, {"DNT" => 1} do |source, req|
      _(req.url).must_equal "http://example.net/streaming"
      _(req.get_args[0]).must_equal({ query: {:chuck => "norris"},
                                      head: {"DNT" => 1,
                                             "Cache-Control" => "no-cache",
                                             "Accept" => "text/event-stream"},
                                      keepalive: true })
      EM.stop
    end
  end

  it "connect and error if status != 200" do
    start_source do |source, req|
      source.error do |error|
        _(error).must_equal "Unexpected response status 400"
        _(source.ready_state).must_equal EM::EventSource::CLOSED
        EM.stop
      end
      source.open { assert false }
      req.call_headers(create_response_headers "400")
    end
  end

  it "connect and error if the content-type doens't match text/event-stream" do
    start_source do |source, req|
      source.error do |error|
        _(error).must_equal "The content-type 'text/plop' is not text/event-stream"
        _(source.ready_state).must_equal EM::EventSource::CLOSED
        EM.stop
      end
      req.call_headers(create_response_headers "200", "text/plop")
    end
  end

  it "connect and error if the content-type is not set" do
    start_source do |source, req|
      source.error do |error|
        _(error).must_equal "The content-type '' is not text/event-stream"
        EM.stop
      end
      req.call_headers(create_response_headers "200", "", "HEADER_NAME" => "BAD")
    end
  end

  it "connect without error with 200 and good content-type" do
    start_source do |source, req|
      source.error do
        assert false
      end
      source.open do
        _(source.ready_state).must_equal EM::EventSource::OPEN
        assert true
        EM.stop
      end
      req.call_headers(create_response_headers "200", "text/event-stream; charset=utf-8")
    end
  end

  it "follow redirect" do
    start_source do |source, req|
      source.error do
        assert false
      end
      source.open do
        assert true
        EM.stop
      end
      req.call_headers(create_response_headers "302", "", {'LOCATION' => "http://example.com/streaming2"})
      assert req.closed
      req.call_callback
      req2 = source.instance_variable_get "@req"
      refute_same(req2, req)
      _(req2.get_args[0]).must_equal({ head: { "Accept" => "text/event-stream",
                                            "Cache-Control" => "no-cache" },
                                    query: {},
                                    keepalive: true })
      req.call_headers(create_response_headers "200", "text/event-stream; charset=utf-8")
      _(source.url).must_equal "http://example.com/streaming2"
    end
  end

  {"LF" => "\n", "CRLF" => "\r\n"}.each do |eol_desc, eol|
    context "with #{eol_desc} EOL" do
      it "connect and handle message" do
        start_source do |source, req|
          source.message do |message|
            _(message).must_equal "hello world"
            source.close
            EM.stop
          end
          req.stream_data("data: hello world#{eol}#{eol}")
        end
      end

      it "handle multiple messages" do
        start_source do |source, req|
          source.message do |message|
            _(message).must_equal "hello world\nplop"
            source.close
            EM.stop
          end
          req.stream_data("data: hello world#{eol}data:plop#{eol}#{eol}")
        end
      end

      it "ignore empty message" do
        start_source do |source, req|
          source.message do |message|
            _(message).must_equal "hello world"
            EM.stop
          end
          req.stream_data(":#{eol}#{eol}")
          req.stream_data("data: hello world#{eol}#{eol}")
        end
      end

      it "removes characters that are not UTF-8" do
        start_source do |source, req|
          source.message do |message|
            _(message).must_equal "foo?bar"
            source.close
            EM.stop
          end
          req.stream_data("data: foo\xE4bar#{eol}#{eol}")
        end
      end

      it "handle event name" do
        start_source do |source, req|
          source.on "plop" do |message|
            _(message).must_equal "hello world"
            source.close
            EM.stop
          end
          req.stream_data("data: hello world#{eol}event:plop#{eol}#{eol}")
        end
      end

      it "reconnect after error with last-event-id" do
        start_source do |source, req|
          req.call_headers(create_response_headers "200", "text/event-stream; charset=utf-8")
          req.stream_data("id: roger#{eol}#{eol}")
          source.error do |error|
            _(error).must_equal "Connection lost. Reconnecting."
            _(source.ready_state).must_equal EM::EventSource::CONNECTING
            EM.add_timer(4) do
              req2 = source.instance_variable_get "@req"
              refute_same(req2, req)
              _(source.last_event_id).must_equal "roger"
              _(req2.get_args[0]).must_equal({ head: { "Last-Event-Id" => "roger",
                                                       "Accept" => "text/event-stream",
                                                       "Cache-Control" => "no-cache" },
                                               query: {},
                                               keepalive: true })
              EM.stop
            end
          end
          req.call_errback
        end
      end

      it "reconnect after callback with last-event-id" do
        start_source do |source, req|
          req.call_headers(create_response_headers "200", "text/event-stream; charset=utf-8")
          req.stream_data("id: roger#{eol}#{eol}")
          source.error do |error|
            _(error).must_equal "Connection lost. Reconnecting."
            _(source.ready_state).must_equal EM::EventSource::CONNECTING
            EM.add_timer(4) do
              req2 = source.instance_variable_get "@req"
              refute_same(req2, req)
              _(source.last_event_id).must_equal "roger"
              _(req2.get_args[0]).must_equal({ head: { "Last-Event-Id" => "roger",
                                                       "Accept" => "text/event-stream",
                                                       "Cache-Control" => "no-cache" },
                                               query: {},
                                               keepalive: true })
              EM.stop
            end
          end
          req.call_callback
        end
      end

      it "handle retry event" do
        start_source do |source ,req|
          req.stream_data("retry: plop#{eol}#{eol}")
          _(source.retry).must_equal 3
          req.stream_data("retry: 45plop#{eol}#{eol}")
          _(source.retry).must_equal 3
          req.stream_data("retry: 1000#{eol}#{eol}")
          _(source.retry).must_equal 1
          req.stream_data("retry: 400#{eol}#{eol}")
          _(source.retry).must_equal 0
          EM.stop
        end
      end
    end
  end

  it "add connection middlewares" do
    start_source do |source ,req|
      proc = Proc.new {}
      source.use "oup", "la", "boom", &proc
      source.close
      source.start
      req2 = source.instance_variable_get "@req"
      _(req2.middlewares).must_equal [["oup", "la", "boom", proc]]
      EM.stop
    end

    start_source do |source ,req|
      proc = Proc.new {}
      source.use "oup", &proc
      source.close
      source.start
      req2 = source.instance_variable_get "@req"
      _(req2.middlewares).must_equal [["oup", proc]]
      EM.stop
    end

    start_source do |source ,req|
      proc = Proc.new {}
      source.use "oup", nil, nil, &proc
      source.close
      source.start
      req2 = source.instance_variable_get "@req"
      _(req2.middlewares).must_equal [["oup", nil, nil, proc]]
      EM.stop
    end

    start_source do |source ,req|
      source.use "oup"
      source.close
      source.start
      req2 = source.instance_variable_get "@req"
      _(req2.middlewares).must_equal [["oup", nil]]
      EM.stop
    end

    start_source do |source ,req|
      source.use "oup", "la", "boom"
      source.close
      source.start
      req2 = source.instance_variable_get "@req"
      _(req2.middlewares).must_equal [["oup", "la", "boom", nil]]
      EM.stop
    end

    start_source do |source ,req|
      source.use("oup", "la", "boom") do
        true
      end
      source.close
      source.start
      req2 = source.instance_variable_get "@req"
      _(req2.middlewares.size).must_equal 1
      _(req2.middlewares[0].size).must_equal 4
      _(req2.middlewares[0][0..2]).must_equal ["oup", "la", "boom"]
      _(req2.middlewares[0][3].class).must_equal Proc
      EM.stop
    end
  end

  it "keeps connection middlewares calls the same" do
    start_source do |source ,req|
      source.use "Keep", "me", "equal"
      source.close
      source.start
      req2 = source.instance_variable_get "@req"
      _(req2.middlewares).must_equal [["Keep", "me", "equal", nil]]
      source.close
      source.start
      req2 = source.instance_variable_get "@req"
      _(req2.middlewares).must_equal [["Keep", "me", "equal", nil]]
      EM.stop
    end
  end

  it "allows to set the inactivity_timeout" do
    EM.run do
      source = EventMachine::EventSource.new("")
      source.inactivity_timeout = 0
      source.start
      req = source.instance_variable_get "@req"
      _(req.opts[:inactivity_timeout]).must_equal 0
      EM.stop
    end
  end

  it "doesn't fail when trying to close not yet started source" do
    EventMachine::EventSource.new("").close
  end

end


lib/em-eventsource.rb
# encoding: utf-8

require "eventmachine"
require "em-http-request"

module EventMachine
  # EventSource
  # dev.w3.org/html5/eventsource/
  class EventSource
    # Get API url
    attr_reader :url
    # Get ready state
    attr_reader :ready_state
    # Get current retry value (in seconds)
    attr_reader :retry
    # Override retry value (in seconds)
    attr_writer :retry
    # Get value of last event id
    attr_reader :last_event_id
    # Get the inactivity timeout
    attr_reader :inactivity_timeout
    # Set the inactivity timeout
    attr_writer :inactivity_timeout
    # Ready state
    # The connection has not yet been established, or it was closed and the user agent is reconnecting.
    CONNECTING = 0
    # The user agent has an open connection and is dispatching events as it receives them.
    OPEN       = 1
    # The connection is not open, and the user agent is not trying to reconnect. Either there was a fatal error or the close() method was invoked.
    CLOSED     = 2

    # Create a new stream
    #
    # url - the url as string
    # query - the query string as hash
    # headers - the headers for the request as hash
    def initialize(url, query={}, headers={})
      @conn = nil
      @url = url
      @query = query
      @headers = headers
      @ready_state = CLOSED

      @last_event_id = nil
      @retry = 3 # seconds
      @inactivity_timeout = 60 # seconds

      @opens = []
      @errors = []
      @messages = []
      @on = {}
      @middlewares = []
    end

    # Add open event handler
    #
    # Returns nothing
    def open(&block)
      @opens << block
    end

    # Add a specific event handler
    #
    # name - name of event
    #
    # Returns nothing
    def on(name, &block)
      @on[name] ||= []
      @on[name] << block
    end

    # Add message event handler
    #
    # Returns nothing
    def message(&block)
      @messages << block
    end

    # Add error event handler
    #
    # Returns nothing
    def error(&block)
      @errors << block
    end

    # Add a middleware
    #
    # *args - the middleware class
    #
    # Returns nothing
    def use(*args, &block)
      block = args.pop if block.nil? && block_given?
      @middlewares << [args, block]
    end

    # Start subscription
    #
    # Returns nothing
    def start
      @ready_state = CONNECTING
      listen
    end

    # Cancel subscription
    #
    # Returns nothing
    def close
      @ready_state = CLOSED
      @conn.close('requested') if @conn
    end

    protected

    def listen
      @conn, @req = prepare_request
      @req.headers(&method(:handle_headers))
      @req.errback(&method(:handle_reconnect))
      @req.callback(&method(:handle_reconnect))
      buffer = ""
      @req.stream do |chunk|
        unless chunk.valid_encoding?
          chunk.scrub!('?')
        end
        buffer += chunk
        while index = buffer.index(/\r\n\r\n|\n\n/)
          stream = buffer.slice!(0..index)
          handle_stream(stream)
        end
      end
    end

    def handle_reconnect(*args)
      return if @ready_state == CLOSED

      if @req.response_header.redirection?
        @url = @req.response_header.location
        listen
        return
      end

      @ready_state = CONNECTING
      @errors.each { |error| error.call("Connection lost. Reconnecting.") }
      EM.add_timer(@retry) do
        listen
      end
    end

    def handle_headers(headers)
      if headers.redirection?
        @req.close
        return
      end
      if headers.status != 200
        close
        @errors.each { |error| error.call("Unexpected response status #{headers.status}") }
        return
      end
      if /^text\/event-stream/.match headers['CONTENT_TYPE']
        @ready_state = OPEN
        @opens.each { |open| open.call }
      else
        close
        @errors.each { |error| error.call("The content-type '#{headers['CONTENT_TYPE']}' is not text/event-stream") }
      end
    end

    def handle_stream(stream)
      data = ""
      name = nil
      stream.split(/\r?\n/).each do |part|
        /^data:(.+)$/.match(part) do |m|
          data += m[1].strip
          data += "\n"
        end
        /^id:(.+)$/.match(part) do |m|
          @last_event_id = m[1].strip
        end
        /^event:(.+)$/.match(part) do |m|
          name = m[1].strip
        end
        /^retry:(.+)$/.match(part) do |m|
          if m[1].strip! =~ /^[0-9]+$/
            @retry = m[1].to_i / 1000
          end
        end
      end
      return if data.empty?
      data.chomp!
      if name.nil?
        @messages.each { |message| message.call(data) }
      else
        @on[name].each { |message| message.call(data) } unless @on[name].nil?
      end
    end

    def prepare_request
      conn = EM::HttpRequest.new(@url, {tls: {verify_peer: true}, inactivity_timeout: @inactivity_timeout})
      @middlewares.each do |middleware, block|
        if block
          conn.use(*middleware, &block)
        else
          conn.use(*middleware)
        end
      end
      headers = @headers.merge({'Cache-Control' => 'no-cache', 'Accept' => 'text/event-stream'})
      headers.merge!({'Last-Event-Id' => @last_event_id }) unless @last_event_id.nil?
      [conn, conn.get({ query: @query,
                        head: headers,
                        keepalive: true })]
    end
  end
end
